File: Top 100 Liked Questions/test_2/test_2.dart
void main(List<String> args) {}

ListNode? addTwoNumbers(ListNode? l1, ListNode? l2) {
  int carry = 0;
  ListNode? head = new ListNode(0);
  ListNode? p = l1, q = l2, curr = head;
  while (p != null || q != null) {
    int x = (p != null) ? p.val : 0;
    int y = (q != null) ? q.val : 0;
    int sum = carry + x + y;
    carry = sum ~/ 10;
    curr!.next = new ListNode(sum % 10);
    curr = curr.next!;
    p = p?.next;
    q = q?.next;
  }
  if (carry > 0) {
    curr!.next = new ListNode(carry);
  }
  return head.next!;
}

class ListNode {
  int val;
  ListNode? next;
  ListNode([this.val = 0, this.next]);
}

File: Top 100 Liked Questions/test_2/test_2.js
function ListNode(val) {
  this.val = val;
  this.next = null;
}
var addTwoNumbers = function (l1, l2) {
  let carry = 0;
  let head = new ListNode(0);
  let current = head;
  while (l1 || l2 || carry) {
    let sum = (l1 ? l1.val : 0) + (l2 ? l2.val : 0) + carry;
    carry = Math.floor(sum / 10);
    current.next = new ListNode(sum % 10);
    current = current.next;
    l1 = l1 ? l1.next : null;
    l2 = l2 ? l2.next : null;
  }
  return head.next;
};

console.log(
  addTwoNumbers(
    ListNode(2, ListNode(4, ListNode(3, null))),
    ListNode(5, ListNode(6, ListNode(4, null)))
  )
);
